modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/MonadDefer.kt

```diff
@@ -36,6 +38,11 @@ interface MonadDefer<F> : MonadThrow<F>, Bracket<F, Throwable> {
   fun <A> delayOrRaise(f: () -> Either<Throwable, A>): Kind<F, A> =
     defer { f().fold({ raiseError<A>(it) }, { just(it) }) }
 
+  /**
+   * Creates a [Ref] to purely manage mutable state, initialized by the function [f]
+   */
+  fun <A> ref(f: () -> A): Kind<F, Ref<F, A>> = Ref(this, f)
```

Comment:
```suggestion
  fun <A> Ref(f: () -> A): Kind<F, Ref<F, A>> = Ref(this, f)
```

Should we fake constructors? This a common pattern that they use throughout the Kotlin lib as well.

Reply:
I'd rather not for conflict purposes.

Reply:
Conflict with? This is only accessible in scoped areas `IO.concurrent().run { }`, `IO.concurrent().fx.concurrent { }`, `IO.fx { }`.

Reply:
With the constructor I mean.

Reply:
Different overrides tho. But still, I don't like it.

Reply:
> Ref(), Semaphore(5), Promise(), MVar(), MVar(5)

I'd still like to discuss this because I am really in favor of doing them with capital letter but if we all agree to them as regular functions I can live with that but we should be consistent. And/or properly define when to do this and when not, with a consistent API for 1.0 in mind.

* We currently do this for [State](https://github.com/arrow-kt/arrow/blob/c2de52c6f5850c6b67d54366a43203cebb0503a3/modules/core/arrow-extras-data/src/main/kotlin/arrow/data/State.kt#L50) to provide a Kotlin idiomatic `State` constructor over `StateT`. I got stuck improving the API further.
* Kotlinx does this also for [Runnable](https://github.com/Kotlin/kotlinx.coroutines/blob/34287743c5cb2ae6cdd0aab09a4b42403ae82ef7/kotlinx-coroutines-core/common/src/Runnable.common.kt#L21) and I've seen PRs doing it for [Semaphore](https://github.com/Kotlin/kotlinx.coroutines/blob/2cbca1acf3f6be5c68339d1b33050e4396854777/kotlinx-coroutines-core/common/src/sync/Semaphore.kt#L25)

Given that the `fx` DSL is going to be one of our selling point AFAIK, I am in favor of giving it the semantics of a constructor there.
In the DSL you'll be able to work in a way you would with KEEP-87, where all instances can be injected by us. Even including syntax that has a default value for `CoroutineContext`.

```
fx.concurrent {
   val (fiber) = IO { "Hello from dispatchers().default() }.fork()
   val (timeOut) = (fiber race sleep(100.miliseconds))
   val (semaphore) = Semaphore(5)
   val (ref) = Ref(1)
   val (parResult) = parMapN(IO { 1 }, IO { 1 }) { a, b -> a + b }
}
```

Reply:
We do `State` because there's no `State` constructor, it's an alias for `StateT<ForId, S, A>` with the value partially applied.

In this case we have a constructor, and the function that'll only be available in MonadDefer/Concurrent contexts. They have different overloads, so that's a positive. By the time KEEP87 lands, the new function will be obsolete anyway.

I don't have strong feelings either way, if you want to change it go for it.

Reply:
> By the time KEEP87 lands, the new function will be obsolete anyway.

Which means we can remove the alias in `Concurrent` and people's code will remain the same, right?

Reply:
fair, fair :D PR?